src/mumble/PipeWire.cpp

```diff
@@ -0,0 +1,437 @@
+			 audioDevice(QStringLiteral("7.1"), QString::number(8)) };
+}
+
+void PipeWireOutputRegistrar::setDeviceChoice(const QVariant &choice, Settings &settings) {
+	settings.qsPipeWireOutput = choice.toString();
+}
+
+bool PipeWireOutputRegistrar::usesOutputDelay() const {
+	return false;
+}
+
+void PipeWireInit::initialize() {
+	pws.reset(new PipeWireSystem());
+
+	pws->m_lock.lock();
+	pws->m_waiter.wait(&pws->m_lock, 1000);
+	pws->m_lock.unlock();
+
+	if (pws->m_ok) {
+		inputRegistrar.reset(new PipeWireInputRegistrar());
+		outputRegistrar.reset(new PipeWireOutputRegistrar());
+	} else {
+		pws.reset();
+	}
+}
+
+void PipeWireInit::destroy() {
+	inputRegistrar.reset();
+	outputRegistrar.reset();
+	pws.reset();
+}
+
+static PipeWireInit pwi;
+
+PipeWireSystem::PipeWireSystem() : m_ok(false), m_users(0) {
+	const QStringList names{ "libpipewire.so", "libpipewire-0.3.so" };
+
+	for (const auto &name : names) {
+		m_lib.setFileName(name);
+		if (m_lib.load()) {
+			break;
+		}
+	}
+
+	if (!m_lib.isLoaded()) {
+		return;
+	}
+
+	RESOLVE(pw_get_library_version);
+	RESOLVE(pw_init);
+	RESOLVE(pw_deinit);
+	RESOLVE(pw_loop_new);
+	RESOLVE(pw_loop_destroy);
+	RESOLVE(pw_thread_loop_new_full);
+	RESOLVE(pw_thread_loop_destroy);
+	RESOLVE(pw_thread_loop_start);
+	RESOLVE(pw_thread_loop_stop);
+	RESOLVE(pw_properties_new);
+	RESOLVE(pw_stream_new_simple);
+	RESOLVE(pw_stream_destroy);
+	RESOLVE(pw_stream_connect);
+	RESOLVE(pw_stream_dequeue_buffer);
+	RESOLVE(pw_stream_queue_buffer);
+
+	qInfo("PipeWire %s from %s", pw_get_library_version(), qPrintable(m_lib.fileName()));
+
+	pw_init(nullptr, nullptr);
+
+	m_ok = true;
+}
+
+PipeWireSystem::~PipeWireSystem() {
+	if (m_ok) {
+		pw_deinit();
+	}
+}
+
+PipeWireEngine::PipeWireEngine(const char *category, void (*callback)(void *data))
+	: m_ok(false), m_loop(nullptr), m_stream(nullptr), m_thread(nullptr) {
+	if (!pws->isOk()) {
+		return;
+	}
+
+	m_loop = pws->pw_loop_new(nullptr);
+	if (!m_loop) {
+		return;
+	}
+
+	auto props = pws->pw_properties_new(PW_KEY_MEDIA_NAME, "Mumble", PW_KEY_MEDIA_CATEGORY, category, PW_KEY_MEDIA_TYPE,
+										"Audio", PW_KEY_MEDIA_ROLE, "Communication", nullptr);
+
+	m_events          = std::make_unique< pw_stream_events >();
+	m_events->version = PW_VERSION_STREAM_EVENTS;
+	m_events->process = callback;
+
+	m_stream = pws->pw_stream_new_simple(m_loop, category, props, m_events.get(), this);
+	if (!m_stream) {
+		return;
+	}
+
+	m_thread = pws->pw_thread_loop_new_full(m_loop, category, nullptr);
+	if (m_thread) {
+		m_ok = true;
+	}
+}
+
+PipeWireEngine::~PipeWireEngine() {
+	if (!pws->isOk()) {
+		return;
+	}
+
+	if (m_thread) {
+		pws->pw_thread_loop_destroy(m_thread);
+	}
+
+	if (m_stream) {
+		pws->pw_stream_destroy(m_stream);
+	}
+
+	if (m_loop) {
+		pws->pw_loop_destroy(m_loop);
+	}
+}
+
+bool PipeWireEngine::connect(const uint8_t direction, const uint32_t *channels, const uint8_t nChannels) {
+	uint8_t buffer[1024];
```

Comment:
Why 1024?

Reply:
It's an arbitrary size used in PipeWire tutorials, e.g.: https://docs.pipewire.org/md__builds_pipewire_pipewire_doc_tutorial4.html